src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSTree.java

```
@@ -0,0 +1,322 @@
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has non-existent child " + child);
+                }
+                if (!tree.get(child).getParent().equals(id)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has child " + child + ", whose parent is " + tree.get(child).getParent() + " instead of " + id);
+                }
+            }
+            final String parent = n.getParent();
+            if (parent != null) {
+                if (!tree.containsKey(parent)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has non-existent parent " + parent);
+                }
+                if (!tree.get(parent).getChildren().contains(id)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has parent " + parent + ", which does not have child " + id);
+                }
+            }
+        }
+
+        //Check disconnected sets of nodes using a traversal from the root
+        final Set<String> reachable = traverse();
+        if (reachable.size() != tree.size()) {
+            final Set<String> unreachable = tree.keySet();
+            unreachable.removeAll(reachable);
+            final String nodesList = getAbbreviatedNodeListString(unreachable);
+            throw new UserException.BadInput("Malformed tree detected. Tree is disconnected. " + unreachable.size() + " of " + tree.size() + " nodes were unreachable: " + nodesList);
+        }
+    }
+
+    /**
+     * Find all nodes reachable from the root
+     */
+    private Set<String> traverse() {
+        final Queue<String> queue = new LinkedList<>();
+        final Set<String> visited = new HashSet<>(tree.size());
+        queue.add(root);
+        while (!queue.isEmpty()) {
+            final String id = queue.poll();
+            if (!visited.contains(id)) { //checked visited in case there are cycles
```

Would it be better to throw an exception here? I'm worried that it could cause a loop in `getPathOf` for some reason the tree actually did contain cycles.